### PR TITLE
Feat/academy/progress tracker

### PR DESCRIPTION
**Pull request by respp**
# Academy Progress Tracker Contract Implementation

## Overview
Implementation of a smart contract to track student progress in an academy setting. The contract allows tracking lesson completion by user and chapter, calculating completion percentages, and emitting events for progress tracking.

## Features Implemented
- **Lesson Tracking**: Track completion of individual lessons by user and chapter
- **Chapter Progress**: Calculate and track chapter completion percentage
- **Event System**: Emit events for both lesson and chapter completion
- **Duplicate Prevention**: Prevent marking the same lesson as completed multiple times
- **Progress Queries**: Functions to query completed lessons and chapter progress

## Technical Details
- **Contract Structure**:
  - `lib.rs`: Main contract implementation
  - `events.rs`: Event definitions and handling
  - `errors.rs`: Custom error types and handling
  - `testutils.rs`: Testing utilities
  - `test.rs`: Unit tests

- **Key Functions**:
  - `mark_lesson_complete`: Mark a lesson as completed
  - `get_completed_lessons`: Get list of completed lessons
  - `get_chapter_completion_percent`: Calculate chapter completion percentage
  - `is_chapter_complete`: Check if chapter is fully completed
  - `set_chapter_lessons`: Configure total lessons in a chapter

- **Events**:
  - `lesson_completed`: Emitted when a lesson is completed
  - `chapter_completed`: Emitted when all lessons in a chapter are completed

## Testing
- Unit tests implemented for all major functionality
- Test coverage for:
  - Lesson completion tracking
  - Chapter progress calculation
  - Event emission
  - Error handling
  - Edge cases (duplicate completions, invalid inputs)

## Documentation
- README.md with:
  - Contract description
  - Building instructions
  - Testing instructions
  - Function documentation
  - Event documentation

## Dependencies
- soroban-sdk
- testutils for testing

## Build Instructions
```bash
stellar contract build --manifest-path ./contracts/academy-progress-tracker/Cargo.toml
```
or
```bash
cd contracts/academy-progress-tracker
stellar contract build
```

## Testing Instructions
```bash
cargo test --features testutils
```
image

## Summary by CodeRabbit

- **New Features**
  - Introduced the Academy Progress Tracker smart contract for tracking student lesson and chapter completion.
  - Added the Academy Graduation NFT contract to mint soulbound NFTs as proof of module completion, integrating with progress and badge trackers.
  - Users can mark lessons as complete, retrieve completed lessons, calculate chapter completion percentages, and check chapter completion status.
  - Events are emitted for lesson and chapter completions.
  - Admins can set the total number of lessons per chapter.
  - Added breadcrumb, label, radio group, dropdown menu, and other UI components for improved navigation and input controls.
  - Added React components for learning paths, lesson views, lesson navigation, knowledge checks, and resource filtering to enhance the learning experience.

- **Documentation**
  - Added READMEs for Academy Progress Tracker and Academy Graduation NFT contracts with detailed usage and setup instructions.
  - Added comprehensive test suite documentation for the AuthController contract.

- **Tests**
  - Added extensive unit and integration tests for Academy Graduation NFT and AuthController contracts covering initialization, minting, authentication, and event emission.
  - Added basic tests and utilities for the Academy Progress Tracker contract.

- **Chores**
  - Updated project configurations to include new contract modules and dependencies.
  - Reformatted code styling and imports across multiple frontend components for consistency.

### CODE REVIEW

File: apps/contract/contracts/academy-progress-tracker/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        // Check if the lesson is already completed
+        let user_progress = env
+            .storage()
+            .persistent()
+            .get::<DataKey, Vec<u32>>(&DataKey::UserChapterProgress(user.clone(), chapter_id))
+            .unwrap_or(vec![&env]);
+
+        if user_progress.contains(&lesson_id) {
+            return Err(Error::LessonAlreadyCompleted);
+        }
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/contract/contracts/academy-progress-tracker/src/lib.rs:43]: Doesn't matter if the lesson is long/short, all of them will keep the same off-chain reference (hashed metadata and/or off-chain data row ID) to reduce the amount of resources required to keep them in-memory within the contract. This way we standardize the data stored and we keep the size as similar as possible from their siblings.

**Review by AndlerRL — CHANGES_REQUESTED** [review]
LGTM! Nice academy contract :+1: however I found maybe a typo in one of the changes, you can take it a look. I have also left the coderabbitai feedbacks to take care!

After those it will be good to go!

**Comment by Bran18** [comment]
> > ¡Genial! ¡Buen contrato de academia! 👍 Sin embargo, encontré una errata en uno de los cambios; puedes revisarlo. También dejé los comentarios de coderabbitai para que se encarguen.
> > ¡Después de esto estará todo listo para empezar!
> 
> Great Anlder, i'll get to work on that.

Hey @respp  I was about to merge your code, but I noticed that there is a Git conflict that needs to be resolved. Could you please handle this? This way, I won’t have to create a commit myself, which might affect your work. Thank you!

**Comment by AndlerRL** [comment]
Closing this PR in favor of PR #512, which already has the requested changes (and merged) and with the latest updates and synced with develop.
